Added kinematics plugin that uses Levenberg-Marquardt method

fsuarez6 opened the pull request:
Here I propose to include a new kinematics plugin: `lma_kinematics_plugin`

This could be a good alternative to the default Newton-Raphson plugin (aka. `kdl_kinematics_plugin`) because is faster and more effective.

I've performed a test with 100.000 valid goals (poses) for [this](http://www.azorobotics.com/images/equipments/EquipmentImage_251.jpg) 6 DoF robot and these are the results:

| Plugin | Solved (%) | Mean Time (ms.) |
| --- | :-: | :-: |
| `ikfast61` | 99,6 | 1,8 |
| `kdl_kinematics_plugin` | 81.3 | 9,2 |
| `lma_kinematics_plugin` | 96,2 | 1,6 |

All of them check that the solution is within the joint limits defined in the URDF.

Will perform tests against other robots.

fsuarez6 committed:
Added kinematics plugin that uses Levenberg-Marquardt method

davetcoleman commented on planning/lma_kinematics_plugin/include/moveit/lma_kinematics_plugin/chainiksolver_pos_lma_jl_mimic.hpp:
```diff
@@ -0,0 +1,102 @@
+// linear relationship to that of another joint.
+// Copyright  (C)  2013  Sachin Chitta, Willow Garage
+
+#ifndef KDLChainIkSolverPos_LMA_JL_Mimic_HPP
```
I believe this should follow the ROS package guidelines of being pre-pended with the package name (all caps): lma_kinematics_plugin__

Also, it'd be nicer if the file name ended with *.h since everything else in MoveIt is.

davetcoleman commented on planning/lma_kinematics_plugin/include/moveit/lma_kinematics_plugin/chainiksolver_pos_lma_jl_mimic.hpp:
```diff
@@ -0,0 +1,102 @@
+
+/**
+ * Implementation of a general inverse position kinematics
+ * algorithm based on Newton-Raphson iterations to calculate the
```
Not 'Newton-Raphson'

davetcoleman commented on planning/lma_kinematics_plugin/include/moveit/lma_kinematics_plugin/chainiksolver_pos_lma_jl_mimic.hpp:
```diff
@@ -0,0 +1,102 @@
+   * @param q_min the minimum joint positions
+   * @param fksolver a forward position kinematics solver
+   * @param iksolver an inverse velocity kinematics solver
+   * @param maxiter the maximum Newton-Raphson iterations,
```
Not 'Newton-Raphson'

davetcoleman commented on planning/lma_kinematics_plugin/include/moveit/lma_kinematics_plugin/chainiksolver_pos_lma_jl_mimic.hpp:
```diff
@@ -0,0 +1,102 @@
+
+}
+
+#endif
```
Does this file belong in the KDL library itself? Is it copied from there?

davetcoleman commented on planning/lma_kinematics_plugin/include/moveit/lma_kinematics_plugin/chainiksolver_pos_lma_jl_mimic.hpp:
```diff
@@ -0,0 +1,102 @@
+
+  bool setMimicJoints(const std::vector<lma_kinematics_plugin::JointMimic>& mimic_joints);
+
+private:
```
MoveIt! style has member variables sufficed with "_", not sure if that applies here.

davetcoleman commented:
Great work, especially with benchmarking the results!

It seems you duplicated a lot of code from the `kdl_kinematics_plugin`? Since this plugin is also kdl-based, perhaps we should combine this plugin with the kdl plugin and just have a rosparam or other setting that switches between the two. This makes sense in terms of naming, since they are both kdl plugins really. I'm not sure how much over lap they have in code, though.

Thanks for this contribution!

fsuarez6 commented:
Yes. I duplicated a lot from `kdl_kinematics_plugin`. I agree with you about the rosparam approach and think that will be a better and feasible solution. Will work on it to see how difficult it is.

davetcoleman commented:
Great! I'd add a setting to the kinematic config file, [like with the pr2](https://github.com/davetcoleman/moveit_pr2/blob/hydro-devel/pr2_moveit_config/config/kinematics.yaml)

sachinchitta commented:
I would say no to adding this into the kdl plugin just because it will make that plugin hard to maintain and debug. I agree that there are ways to extend the KDL kinematics plugin (e.g. different solvers like the one added here) but that would require a different architecture. In the meantime, this works as a different plugin if all the changes @davetcoleman mentions are included.

alainsanguinetti commented:
Sounds like a nice PR, are there any reasons why it hasn't been merged in the end ? If I make the proposed changes, would that be enough ?

davetcoleman commented:
the requested improvements were never made so it was not merged in

v4hn commented:
given tracIK is pretty popular right now, I'm not sure this is still interesting.
@alainsanguinetti: if you want to see this merged, fork `fsuarez6:hydro-devel`,
get it to work with indigo-devel or jade-devel, address the comments in this request
and open a new pull-request where we can discuss whether anything else needs to be changed
before merging.

alainsanguinetti commented:
@v4hn is tracIK integrated in MoveIt! ? I am thinking about integrating this one but I need to check with my boss if I am allowed to spend time on it.

v4hn commented:
On Tue, Jul 05, 2016 at 02:34:52AM -0700, Alain Sanguinetti wrote:

> @v4hn is tracIK integrated in MoveIt! ?

https://groups.google.com/forum/#!topic/moveit-users/hsrHK-JKEnY

fsuarez6 commented:
@alainsanguinetti have you started with the changes proposed by @davetcoleman ?

Even though TRAC-IK seems to be better that the current KDL plugin and the LMA proposed in this pull request, I think it would be nice to have alternatives. You may find yourself working with a robot where LMA behaves the best.

fsuarez6 commented:
@alainsanguinetti you can test the LMA plugin using [this repository](https://github.com/fsuarez6/lma_kinematics_plugin). That way you don't need to modify your MoveIt installation.

alainsanguinetti commented:
@fsuarez6 Hello, I haven't started working on it but I'm still interested in this plugin :+1: Thanks for the repo, I'll probably fork it someday to test it and create an Indigo version.

fsuarez6 commented:
@alainsanguinetti I'll work on this pull request so that we can have the LMA plugin as part of MoveIt.

davetcoleman commented:
+1 to cleaning it up, thanks!

fsuarez6 committed:
Made changes suggested by @davetcoleman:
- Modified the include guards to comply ROS packages guidelines
- Changed header file extensions from .hpp to .h
- Fixed typos in the documentation mentioning Newton-Raphson

fsuarez6 commented:
Made changes suggested by @davetcoleman:
- Modified the include guards to comply ROS packages guidelines
- Changed header file extensions from .hpp to .h
- Fixed typos in the documentation mentioning Newton-Raphson

> I would say no to adding this into the kdl plugin just because it will make that plugin hard to maintain and debug. I agree that there are ways to extend the KDL kinematics plugin (e.g. different solvers like the one added here) but that would require a different architecture. In the meantime, this works as a different plugin

@davetcoleman taking into account @sachinchitta comment (above), I think the plugin is ready to be merged. If there is a change in MoveIt's architecture in the future, we could put together all the KDL plugins.

v4hn commented on planning/lma_kinematics_plugin/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+target_link_libraries(${MOVEIT_LIB_NAME} moveit_rdf_loader ${catkin_LIBRARIES})
+
+install(TARGETS ${MOVEIT_LIB_NAME} LIBRARY DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION})
+install(DIRECTORY include/ DESTINATION include)
```
please use `${CATKIN_PACKAGE_INCLUDE_DESTINATION}` as destination

v4hn commented:
I'm pretty sure we are not developing for hydro anymore, but the requests targets hydro-devel.
Could you add a new request for indigo-devel?

fsuarez6 committed:
Changed hard-coded installation destination for headers

fsuarez6 commented:
Moved to #723
